Feature/create categeory tv show screen

monaayman123 opened the pull request:
## Description
- This PR introduces the Category screen and CategoryDetails screen implementation.
It includes displaying genres as chips and loading the corresponding movies/TV shows when a genre is selected.
Also includes handling loading state, error state, and retry functionality and add pagination support for category movies details

---

## Changes Made
- Created Category screen with list of genres as selectable chips.
- Implemented CategoryDetails screen to display movies or TV shows by selected genre.
- Integrated GetMoviesByGenreIdUseCase and newly for fetching media by genre.

- Added logic in CategoriesDetailsViewModel to:
- Load initial genre from navigation arguments.
- Update UI when a different genre is selected.
- Handle loading and error states with retry option.

- Added CategoriesDetailsUiState with:
- IsLoading flag for showing progress indicator.
- ErrorUiState for handling network and unknown errors.
- Movies list to display fetched media.
- Updated UI mappers (toMovieUiState) for displaying data properly.
- Implemented error handling for NetworkException and other exceptions.
- Implement Paging 3 integration with Pager and PagingSource
- Load movies by genre with infinite scrolling
- Handle load states (loading, error, empty)
- Preserve UI state when switching genres

---

## Screenshots (if applicable)
https://github.com/user-attachments/assets/c3f19ad6-8425-46d6-8e26-da2b35d21574

---

## Checklist
- [x] My code follows the code style of this project
- [x] Changes have been tested manually and verified.
- [x] PR includes at most one single feature.

monaayman123 committed:
refactor: delete unneeded file

monaayman123 committed:
refactor: modify padding in the top

monaayman123 committed:
feature: create main category screen

monaayman123 committed:
feature: create CategoriesUiState.kt

monaayman123 committed:
feature: create CategoriesUiEffect.kt

monaayman123 committed:
feature: create CategoriesViewModel.kt

monaayman123 committed:
feature: add CategoriesDetailsViewModel.kt

monaayman123 committed:
feature: add CategoriesDetailsInteractionListener.kt

monaayman123 committed:
feature: add CategoriesDetailsInteractionListener.kt

monaayman123 committed:
feature: add CategoriesDetailsScreen.kt

monaayman123 committed:
feature: add  CategoriesDetailsScreen.kt

monaayman123 committed:
feature: add  CategoriesDetailsUiEffect.kt

monaayman123 committed:
feature: add  CategoriesDetailsUiState.kt

monaayman123 committed:
feature: add  CategoriesDetailsUiStateMapper.kt

monaayman123 committed:
feature: add  CategoriesDetailsViewModel.kt

monaayman123 committed:
feature: add  CategoriesInteractionListener.kt

monaayman123 committed:
feature: add  GenreUiModel.kt

monaayman123 committed:
feature: add  LocalGenres.kt

monaayman123 committed:
feature: add categoriesDetailsScreenRoute()

monaayman123 committed:
feature: add CategoriesDetails()

monaayman123 committed:
feature: add provideGetMoviesByGenreId()

monaayman123 committed:
feature: add provideCategoriesMoviesDetailsArgs()

monaayman123 committed:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into feature/create-categeory-screen

# Conflicts:
#	app/src/main/java/com/amsterdam/aflami/di/useCaseModule.kt

monaayman123 committed:
refactor: remove unneeded import

monaayman123 committed:
refactor: formatting file

monaayman123 committed:
refactor: remove unused variables and renaming file

monaayman123 committed:
refactor: remove unused variables

monaayman123 committed:
refactor: remove unused variables

monaayman123 committed:
refactor: remove unused variables and renaming file

monaayman123 committed:
refactor: remove unused variables

monaayman123 committed:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into feature/create-categeory-screen

monaayman123 committed:
refactor: rename file

monaayman123 committed:
refactor: rename file

monaayman123 committed:
Merge remote-tracking branch 'origin/feature/create-categeory-screen' into feature/create-categeory-screen

monaayman123 committed:
refactor: add pagination

monaayman123 committed:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into feature/create-categeory-screen

monaayman123 committed:
refactor: use single Pager flow driven by currentGenre

monaayman123 committed:
refactor: organize them into movie package

monaayman123 committed:
feature: add GetTvShowsByGenreIdUseCase.kt and injected it in useCaseModule.kt

monaayman123 committed:
feature: add CategoriesTvShowsDetailsInteractionListener.kt

monaayman123 committed:
feature: add CategoriesTvShowsDetailsArgs.kt and injected it in viewModelModule.kt

monaayman123 committed:
feature: add CategoriesMoviesDetailsScreenRoute.kt and add it function in NavGraph.kt

monaayman123 committed:
feature: add CategoriesMoviesDetailsScreen.kt

monaayman123 committed:
feature: add CategoriesTvShowsDetailsUiEffect.kt

monaayman123 committed:
feature: add CategoriesTvShowsDetailsUiState.kt

monaayman123 committed:
feature: add CategoriesTvShowsDetailsUiStateMapper.kt

monaayman123 committed:
feature: add CategoriesTvShowsDetailsViewModel.kt

monaayman123 committed:
feature: add DefaultTvShowGenreUiStates.kt

monaayman123 committed:
feature: add TvShowGenreLocalMapper.kt

monaayman123 committed:
feature: add TvShowGenreItemUiState.kt

monaayman123 committed:
feature: organize both into movie package

monaayman123 committed:
feature: add navigation function  to feature: add CategoriesTvShowsDetailsScreen.kt

monaayman123 committed:
feature: implement functions that get data from api

monaayman123 committed:
feature: delete unneeded line

monaayman123 committed:
feature: delete unneeded files

monaayman123 committed:
feature: add GenreMovieUiModel.kt

monaayman123 committed:
feature: add GenreTvShowUiModel.kt

monaayman123 committed:
feature: use GenreTvShowUiModel and GenreMovieUiModel

monaayman123 committed:
feature: use correct name for args

monaayman123 committed:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into feature/create-categeory-tv-show-screen

# Conflicts:
#	app/src/main/java/com/amsterdam/aflami/di/useCaseModule.kt

monaayman123 committed:
feature: convert date toYearString

monaayman123 committed:
refactor : use sealed interface

monaayman123 committed:
refactor : use return direct

monaayman123 committed:
refactor : update name to be NavigateTo

monaayman123 committed:
refactor: remove unnecessary remember from movies collection

monaayman123 committed:
refactor: remove extra comma

monaayman123 committed:
refactor: remove unneeded lines

ThoriaHamdy99 commented on app/src/main/java/com/amsterdam/aflami/di/useCaseModule.kt:
```diff
@@ -167,10 +169,18 @@ object UseCaseModule {
         GetTopRatedMoviesUseCase(movieRepository)
 
+    @Provides
+    fun provideGetMoviesByGenreId(movieRepository: MovieRepository): GetMoviesByGenreIdUseCase =
```
Make the name ends with use case --> provideGetMoviesByGenreIdUseCase
for more readability and consistency.

ThoriaHamdy99 commented on viewModel/src/main/java/com/amsterdam/viewmodel/categoriesDetails/movies/CategoriesMoviesDetailsUiState.kt:
```diff
@@ -0,0 +1,39 @@
+        data object NoNetworkConnection : CategoriesDetailsErrorState
+        data object UnknownError : CategoriesDetailsErrorState
+        companion object {
+            fun toSearchErrorState(exception: Throwable): CategoriesDetailsErrorState =
```
I think you can make the renaming of toSearchErrorState to be toCategoriesDetailsErrorState.

ThoriaHamdy99 commented on viewModel/src/main/java/com/amsterdam/viewmodel/categoriesDetails/movies/CategoriesMoviesDetailsViewModel.kt:
```diff
@@ -0,0 +1,116 @@
+    dispatcherProvider
+), CategoriesMoviesDetailsInteractionListener {
+
+    private val currentGenre = MutableStateFlow<MovieGenre?>(null)
```
we need to handle the state of screen by every thing in the CategoriesTvShowsDetailsUiState not making another mutable to listen on it. As it is not search we need to listen in. It is just  a click state we can make the action after click and change the state and load the movie again

ThoriaHamdy99 commented on viewModel/src/main/java/com/amsterdam/viewmodel/categoriesDetails/movies/CategoriesMoviesDetailsViewModel.kt:
```diff
@@ -0,0 +1,116 @@
+    private val currentGenre = MutableStateFlow<MovieGenre?>(null)
+
+    init {
+        val initialGenre = MovieGenre.valueOf(categoriesMovieDetailsArgs.genre!!)
```
This initialization of getting the args can be in as separate function and this function will be called here

ThoriaHamdy99 commented on viewModel/src/main/java/com/amsterdam/viewmodel/categoriesDetails/movies/CategoriesMoviesDetailsViewModel.kt:
```diff
@@ -0,0 +1,116 @@
+        val initialGenre = MovieGenre.valueOf(categoriesMovieDetailsArgs.genre!!)
+        currentGenre.value = initialGenre
+
+        val pagingFlow = currentGenre
```
Also getting the movies by genre needs to be in a separate function that wrapped with try to execute and we can call it from here

ThoriaHamdy99 commented on viewModel/src/main/java/com/amsterdam/viewmodel/categoriesDetails/movies/CategoriesMoviesDetailsViewModel.kt:
```diff
@@ -0,0 +1,116 @@
+            }
+            .cachedIn(viewModelScope)
+
+        updateState {
```
the update of the state can be also in a separate function that will be called a ftere the data loaded successfully

ThoriaHamdy99 commented on viewModel/src/main/java/com/amsterdam/viewmodel/categoriesDetails/movies/CategoriesMoviesDetailsViewModel.kt:
```diff
@@ -0,0 +1,116 @@
+        }
+    }
+
+    override fun onClickRetryRequest() {
```
on click retry you need to update the status to be is loaded = true and load the movie data by selected genre

ThoriaHamdy99 replied:
instead of using the mutable value

monaayman123 replied:
done

ThoriaHamdy99 commented on viewModel/src/main/java/com/amsterdam/viewmodel/categoriesDetails/tvShow/CategoriesTvShowsDetailsArgs.kt:
```diff
@@ -0,0 +1,13 @@
+
+
+    companion object{
+        const val GENRE = "genreName"
```
Let the naming to be GENRE_NAME instead of GENRE

ThoriaHamdy99 commented on viewModel/src/main/java/com/amsterdam/viewmodel/categoriesDetails/tvShow/CategoriesTvShowsDetailsArgs.kt:
```diff
@@ -0,0 +1,13 @@
+import androidx.lifecycle.SavedStateHandle
+
+class CategoriesTvShowsDetailsArgs(savedStateHandle: SavedStateHandle) {
+    val genre = savedStateHandle.get<String>(GENRE)
```
and the naming of genre to be genreName

ThoriaHamdy99 commented on viewModel/src/main/java/com/amsterdam/viewmodel/categoriesDetails/tvShow/CategoriesTvShowsDetailsViewModel.kt:
The same points in movie view model need to be handled here

monaayman123 committed:
refactor: make the name ends with use case for more readability and consistency

monaayman123 committed:
refactor: renaming of toSearchErrorState to be toCategoriesDetailsErrorState.

monaayman123 committed:
refactor: add onFetchError function

monaayman123 committed:
refactor: renaming of genre to be genreName
and  renaming to be GENRE_NAME instead of GENRE

ahmedmahgoub0 commented on domain/src/main/java/com/amsterdam/domain/useCase/details/GetMoviesByGenreIdUseCase.kt:
```diff
@@ -0,0 +1,13 @@
+import com.amsterdam.entity.Movie
+import com.amsterdam.entity.category.MovieGenre
+
+class GetMoviesByGenreIdUseCase(
```
get movies by genre or genre name but not id
mapping genre to id in repo

ahmedmahgoub0 commented on ui/src/main/java/com/amsterdam/ui/navigation/NavGraph.kt:
```diff
@@ -50,5 +52,7 @@ fun NavGraph(
         listDetailsScreenRoute()
         watchHistoryScreenRoute()
         myRatingScreenRoute()
+        categoriesDetailsScreenRoute()
```
categories of movies for better naming

monaayman123 committed:
refactor: CategoriesTvShowsDetailsViewModel to improve state handling and initialization

- Removed unnecessary MutableStateFlow for currentGenre
- Updated state handling to fully rely on CategoriesTvShowsDetailsUiState
- Added separate function to initialize genre from navigation args
- Moved click handling logic to update state and reload TV shows immediately
- Ensured pagination logic remains integrated with updated state changes

monaayman123 committed:
feature: add loading state handling for genre based TV shows fetch

monaayman123 committed:
refactor: renaming of genre to be genreName

monaayman123 committed:
refactor: renaming selectedGenreName to be selectedGenre

monaayman123 committed:
refactor: move logic into separate function based on UI state
- Moved movie loading logic into a dedicated function.
- Ensured behavior depends on the current UI state for better separation of concerns.
- Extracted UI state update and data fetching into separate functions.
- Made functions depend on current UI state to improve readability and maintainability.

monaayman123 committed:
refactor: renaming GetTvShowsByGenreIdUseCase.kt to be GetTvShowsByGenreUseCase.kt
and GetMoviesByGenreIdUseCase.kt to be GetMoviesByGenreUseCase.kt

monaayman123 committed:
refactor: renaming categoriesDetailsScreenRoute to be categoriesMoviesDetailsScreenRoute

monaayman123 committed:
fix: unify no network error handling for movies and TV shows in categories details screen

monaayman123 committed:
refactor: renaming provideGetMoviesByGenreIdUseCase
to be provideGetMoviesByGenreUseCase
and provideGetTvShowByGenreIdUseCase to be provideGetTvShowByGenreUseCase

monaayman123 committed:
fix: add loadState observer in TV Shows screen to handle network errors

Belal-Alkhatib commented on domain/src/main/java/com/amsterdam/domain/repository/TvShowRepository.kt:
```diff
@@ -24,4 +25,5 @@ interface TvShowRepository {
         seasonNumber: Int,
         episodeNumber: Int
     ): String
+    suspend fun getTvShowsByGenre(tvShowGenres: List<TvShowGenre>, page: Int): List<TvShow>
```
in our app current design, the user selects only one genre at a time. 
based on this, it would be better to simplify the function by accepting a single TvShowGenre instead of a list.

monaayman123 replied:
done

Belal-Alkhatib commented on domain/src/main/java/com/amsterdam/domain/useCase/details/GetMoviesByGenreUseCase.kt:
```diff
@@ -0,0 +1,13 @@
+class GetMoviesByGenreUseCase(
+    private val movieRepository: MovieRepository
+) {
+    suspend operator fun invoke(selectedGenre: MovieGenre, page: Int): List<Movie> {
+        return movieRepository.getMoviesByGenres(listOf(selectedGenre), page)
+    }
```
every time you call `getMoviesByGenres`, you have to wrap the single `selectedGenre `inside a list using` listOf(selectedGenre).` To simplify this, I suggest changing the function to accept **just** **one genre**. Then, at the lowest level (in the data level), you can wrap that single genre into a list before making the API call.

ThoriaHamdy99 replied:
This function is used in another feature mood picker and I am passing a list so I suggest to add another function not changing this function

Belal-Alkhatib commented on remoteDatasource/src/main/java/com/amsterdam/remotedatasource/datasource/TvRemoteDataSourceImpl.kt:
```diff
@@ -77,13 +80,21 @@ class TvRemoteDataSourceImpl @Inject constructor(
+    override suspend fun getTvShowsByGenreIds(
+        genresIds: List<Long>,
+        page: Int
+    ): RemoteTvShowResponse {
+        return responseCall { tvShowsApiService.getTvShowsByGenreIds(genresIds, page) }
+
+    }
```
here where you can do the wrapping you mentioned: instead of requiring callers to pass a list of genre IDs, you can change the higher-level function to accept a single genre ID and then wrap it into a list just before calling this function

monaayman123 replied:
done

Belal-Alkhatib commented on ui/src/main/java/com/amsterdam/ui/screens/categoriesDetails/movies/CategoriesMoviesDetailsScreen.kt:
```diff
@@ -0,0 +1,198 @@
+    val state by viewModel.state.collectAsState()
+    val navController = LocalNavController.current
+    val movies = state.movies.collectAsLazyPagingItems()
+    LaunchedEffect(state.selectedGenre) {
+        movies.refresh()
+    }
```
you can make this update in view model when the user change genre

monaayman123 replied:
done
i removed it

monaayman123 committed:
Fix loading indicator not showing on initial load in CategoriesMoviesDetailsViewModel

- Ensure loading state is properly set before data fetch
- Update UI state immediately when genre changes
- Improve error handling and state updates for better UX

monaayman123 committed:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into feature/create-categeory--screen

Belal-Alkhatib commented on ui/src/main/java/com/amsterdam/ui/screens/categoriesDetails/tvShow/CategoriesTvShowsDetailsScreen.kt:
```diff
@@ -0,0 +1,198 @@
+    val state by viewModel.state.collectAsState()
+    val navController = LocalNavController.current
+    val tvShows = state.tvShows.collectAsLazyPagingItems()
+    LaunchedEffect(state.selectedGenre) {
+        tvShows.refresh()
+    }
```
better to handle refreshing the TV shows inside the ViewModel when the selected genre changes.

monaayman123 replied:
done

Belal-Alkhatib commented on viewModel/src/main/java/com/amsterdam/viewmodel/categories/CategoriesInteractionListener.kt:
```diff
@@ -0,0 +1,12 @@
+
+interface CategoriesInteractionListener {
+    fun onChangeTabOption(tabOption: TabOption)
+    fun onNavigateCategoriesDetailsScreen(genreName: String)
+    fun onNavigateCategoriesTvShowsDetailsScreen(genreName: String)
```
Rename the navigation functions to clearly reflect user interactions, such as:

```
onClickMovieGenreCard(genreName: String)
onClickTvShowGenreCard(genreName: String)
```

This makes it obvious these are triggered by user clicks and aligns with our naming style.

ThoriaHamdy99 replied:
Done

Belal-Alkhatib commented on viewModel/src/main/java/com/amsterdam/viewmodel/categoriesDetails/movies/CategoriesMoviesDetailsInteractionListener.kt:
```diff
@@ -0,0 +1,12 @@
+import com.amsterdam.entity.category.MovieGenre
+
+interface CategoriesMoviesDetailsInteractionListener {
+    fun onBackClicked()
+    fun onMovieCardClicked(movieId: Long)
+    fun onGenreClicked(movieGenre: MovieGenre)
+    fun onClickRetryRequest()
+    fun onPagingLoadStateChanged(loadStates: CombinedLoadStates)
+}
```
please use our naming style: `onClickX()`

monaayman123 replied:
done

Belal-Alkhatib commented on viewModel/src/main/java/com/amsterdam/viewmodel/categoriesDetails/tvShow/CategoriesTvShowsDetailsInteractionListener.kt:
```diff
@@ -0,0 +1,13 @@
+import com.amsterdam.entity.category.TvShowGenre
+
+interface CategoriesTvShowsDetailsInteractionListener {
+    fun onBackClicked()
+    fun onTvShowCardClicked(tvShowId: Long)
+    fun onGenreClicked(tvShowGenre: TvShowGenre)
+    fun onClickRetryRequest()
+    fun onPagingLoadStateChanged(loadStates: CombinedLoadStates)
+}
```
please use our naming style: onClickX

monaayman123 replied:
done

Belal-Alkhatib commented on viewModel/src/main/java/com/amsterdam/viewmodel/categoriesDetails/movies/CategoriesMoviesDetailsViewModel.kt:
```diff
@@ -0,0 +1,121 @@
+    dispatcherProvider
+), CategoriesMoviesDetailsInteractionListener {
+    init {
+        getInitialGenre()
```
i gess you don’t need this method in the init block, just set a default selected genre directly in your UI state’s initial value

monaayman123 replied:
i take it from args

Belal-Alkhatib commented on viewModel/src/main/java/com/amsterdam/viewmodel/categoriesDetails/tvShow/CategoriesTvShowsDetailsViewModel.kt:
```diff
@@ -0,0 +1,122 @@
+    override fun onGenreClicked(tvShowGenre: TvShowGenre) {
+        updateUiStateForSelectedGenre(tvShowGenre)
+        loadTvShowsForSelectedGenre()
+    }
```
right now, every click triggers a new API call, even if the user clicks the same genre again. This causes unnecessary requests. add a simple check to skip the update if the **newly selected genre is the same as the current one.**

monaayman123 replied:
done

Belal-Alkhatib commented on viewModel/src/main/java/com/amsterdam/viewmodel/categoriesDetails/movies/CategoriesMoviesDetailsViewModel.kt:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    override fun onClickRetryRequest() {
+        loadMoviesForSelectedGenre()
```
Right now, every click triggers a new API call, even if the user clicks the same genre again. This causes unnecessary requests. Add a simple check to skip the update if the newly selected genre is the same as the current one.

monaayman123 replied:
when i click on retry  i need a request from api

monaayman123 committed:
refactor: renaming function to be onClickX instead of onXClicked for consistency

monaayman123 committed:
refactor: getTvShowsByGenre and getMovieByGenre
 to accept a single genre instead of a list

- Updated function signature to take a single TvShowGenre
- Removed need to wrap genre in a list at call sites
- Moved wrapping logic to the data layer before API call
- Simplified ViewModel calls when fetching TV shows by selected genre

monaayman123 committed:
refactor: delete unneeded lined

monaayman123 committed:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into feature/create-categeory--screen

monaayman123 committed:
refactor: add check to skip reloading TV shows or movies if the selected genre hasn't changed

eslam-99 reviewed (dismissed):
Nice progress

ThoriaHamdy99 committed:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into feature/create-categeory-tv-show-screen

# Conflicts:
#	app/src/main/java/com/amsterdam/aflami/di/useCaseModule.kt

ThoriaHamdy99 committed:
fix: rename the on click functions of the categories screen.
